auth-data-phone/src/main/java/com/google/android/horologist/auth/data/phone/tokenshare/impl/TokenBundleRepositoryImpl.kt

```
@@ -0,0 +1,73 @@
+
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.core.Serializer
+import com.google.android.horologist.auth.data.phone.ExperimentalHorologistAuthDataPhoneApi
+import com.google.android.horologist.auth.data.phone.tokenshare.TokenBundleRepository
+import com.google.android.horologist.data.WearDataLayerRegistry
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * Default implementation for [TokenBundleRepository].
+ */
+@ExperimentalHorologistAuthDataPhoneApi
+public class TokenBundleRepositoryImpl<TokenBundle>(
+    private val dataStore: DataStore<TokenBundle>
+) : TokenBundleRepository<TokenBundle> {
+
+    override suspend fun update(tokenBundle: TokenBundle) {
+        dataStore.updateData { tokenBundle }
+    }
+
+    public companion object {
+
+        private const val DEFAULT_TOKEN_BUNDLE_KEY = "/horologist_token_bundle"
+
+        /**
+         * Factory method for [TokenBundleRepositoryImpl].
+         *
+         * If multiple [token bundles][TokenBundle] are required to be shared, specify a [key] in
+         * order to identify each one, otherwise the same [default][DEFAULT_TOKEN_BUNDLE_KEY] key
+         * will be used and only a single token bundle will be persisted.
+         */
+        public fun <TokenBundle> create(
+            key: String = DEFAULT_TOKEN_BUNDLE_KEY,
+            applicationContext: Context,
+            coroutineScope: CoroutineScope,
+            serializer: Serializer<TokenBundle>
+        ): TokenBundleRepositoryImpl<TokenBundle> = TokenBundleRepositoryImpl(
+            WearDataLayerRegistry.fromContext(
+                application = applicationContext,
+                coroutineScope = coroutineScope
+            ).protoDataStore(
+                path = buildPath(key),
+                coroutineScope = coroutineScope,
+                serializer = serializer
+            )
+        )
+
+        private fun buildPath(key: String) =
```

What about just documenting and being stricter here? Failing on one or t'other.

---

Not sure I follow, is my understand correct that you are proposing to:
1) improve documentation of the create function above to say that the key should start with a forward slash
2) throw an exception if the key does not start with forward slash

For 1, if the forward slash is not provided, it fails at runtime by the wear data layer with a message about the "path". This repository is proposing the concept of "keys" and not "path" so the the runtime error can be confusing to the user, hence I decided to verify and add the forward slash for the user.

---

ok, then maybe fail if it doesn't match some pattern aaaa_aaaa? Ban key including a /?

---

I'm arguing against Postels' Law. But I don't think I agree with it. I prefer fail fast.

---

Your call, just explaining my opinion.

---

sorry, I'm lost, maybe some code could help

---

One of these two for example.

```
        private fun buildPathFromKey(key: String): String {
            check(!key.contains('/'))
            return "/$key"
        }
```

```
        private fun buildPathStrict(key: String): String {
            check(key.startsWith('/'))
            return key
        }
```